doschema/jsvalidation.py

```diff
@@ -0,0 +1,229 @@
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+import collections
+
+import jsonschema
+
+from doschema.errors import JSONSchemaCompatibilityException, \
+    JSONSchemaDependenciesException
+
+
+class ResolverFactory(object):
+    """Factory class for resolvers."""
+
+    default_resolvers = {
+        'jsonschema': jsonschema.RefResolver
+    }
+
+    def __new__(cls, resolver_name):
+        """"New override for factory.
+
+        :param resolver_name: String with resolver name.
+        """
+        if isinstance(resolver_name, str):
+            return ResolverFactory.default_resolvers[resolver_name]
+
+
+class JSONSchemaValidation(object):
+    """Class for checking compatibility between schemas."""
+
+    COLLECTION_KEYS = frozenset(['allOf', 'anyOf', 'oneOf'])
+
+    def __init__(self, ignore_index=True, resolver=False):
+        """Constructor.
+
+        :param ignore_index: If set to True, which is default, it will ignore
+                             array indexes.
+        :param resolver: Resolver used to retrieve referenced schemas. If not
+                         provided it will use default.
+
+        """
+        self.ignore_index = ignore_index
+        self.fields_types_dict = {}
+        self.schema_name = None
+        if not resolver:
+            self.resolver_class = ResolverFactory('jsonschema')
+        else:
+            self.resolver_class = resolver
+
+    def start(self, schema, schema_name):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param schema_name: Name of schema that is currently processed.
+        """
+        self.schema_name = schema_name
+        self.resolver = self.resolver_class.from_schema(schema)
+
+        self.traverse(schema, ())
+
+        self.schema_name = None
+        self.resolver = None
+
+    def traverse(self, curr_schema, curr_field):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        self.check_special(curr_schema, curr_field)
+
+        while '$ref' in curr_schema:
+            path = curr_schema.get('$ref')
+            curr_schema = self.resolver.resolve(path)[1]
+
+        if 'properties' in curr_schema:
+            self.traverse(
+                curr_schema['properties'],
+                curr_field + ('properties', )
+            )
+            if 'dependencies' in curr_schema:
+                self.traverse_dependencies(
+                    curr_schema,
+                    curr_field + ('dependencies', )
+                )
+        else:
```

Comment:
This code would run on anything, even invalid JSON Schema. It is better to test explicitly what is expected. The last empty else case should raise an exception as it is not something you expected.